### ISSUE TITLE
model.history not save output of all epochs  

### ISSUE THREAD

**Issue by luyuhengCN** [skorch-dev/skorch#1037]
Dear,
I came across a problem when using the history of the skorch model NeuralNetRegressor. I defined a callback 'calculate_metric' and it will calculate some metrics and save the metrics to history at the end of the epoch.  And I also used the 'Checkpoint' callback to save the model to monitor one of the metrics.  I set the max_epoch = 10, but when I printed the 'model.history', I found it only had outputs of 4 epochs. I checked the outputs of each epoch in 'model.history', I found the history only saved the outputs of those epochs when 'Checkpoint' saved. I posted the scripts below for your review.  I hope the 'model.history' could save all the outputs of the 10 epochs. I appreciate if you could provide some solutions to this problem.

skorch version: 0.15.0
python:  3.9.18
pytorch: 1.13

`class calculate_metric(Callback):
    def __init__(self, name, X, y, date_filepath):
        self._name = name 
        self.X = X
        self.y = y
        self.date_filepath = date_filepath
        self.best_score_ = -np.inf

    def on_epoch_end(self, net, dataset_train=None, dataset_valid=None):
        # Assuming your PyTorch model has a forward method that returns predictions
        y_pred = net.predict(self.X)

        # Calculate external metric using predictions and external labels
        ic, icir, CulRet, sr = metric_stats(y_pred, self.y, self.date_filepath)
       
        # Print or log the external metric
        history = net.history
        for name, metric in zip(self._name, [ic, icir, CulRet, sr]):
            history.record(name, float(metric))  # must do this type convert, due to the need of checkpoint_history writer!

        if ic > self.best_score_:
            history.record(self._name[0] + '_best', True)
            self.best_score_ = ic
        else:
            pass
            history.record(self._name[0] + '_best', False)

    model = NeuralNetRegressor(
        DeepMLP(input_size, hidden_sizes, output_size, dropout_prob=dropout_prob),
        max_epochs=num_epochs,
        criterion=nn.MSELoss,
        lr=learning_rate,
        optimizer=optim.Adam,
        batch_size=batch_size,
        train_split=None,
        iterator_train__shuffle=True,  # Shuffle training data on each epoch
        iterator_train__num_workers=8,
        iterator_train__pin_memory=True,
        device=device,
        verbose=0,
        callbacks=[
            (calculate_metric(['val_ic', 'val_icir', 'val_CulRet', 'val_sr'], 
                              validate_factor, validate_target, validate_date_filepath)),
            (calculate_metric(['test_ic', 'test_icir', 'test_CulRet', 'test_sr'], 
                              test_factor, test_target, test_date_filepath)),
            (EarlyStopping(monitor='val_ic', patience=patience, lower_is_better=False)),
            (Checkpoint(monitor='val_ic_best', load_best=True, event_name=None,
                        dirname=f'{train.get_context().get_trial_dir()}/checkpoint',
                        ))
        ]
    )
    history = model.history
    print(history)`

**Comment by BenjaminBossan** [comment]
From my point of view, this is working as intended. If you want to have the checkpoint and the complete history, there are a few options:

1. You could set `load_best=False` and then load the checkpoint once you don't need the full history anymore.
2. You could write a small callback that saves the history at the end of each epoch (using `pickle` or `net.history.to_file` if you want a json).
